# Week 2 Assignment 1 - 'Counter 앱 만들고 파일 분리하기'

SangYoonLee1231 opened the pull request:
###  🙇‍♀️ 인사
- 저번 주 과제 복습 및 이번 주 React 학습으로 인해 오늘(수요일) 첫 PR을 드리게 되었습니다.
- 이번에도 잘 부탁드립니다!
 
<br/>

### ☑ 과제 구현 목표
- 'Click me!' 버튼 클릭 시, Counter가 1씩 증가한다.
- 숫자 버튼 클릭 시, Counter가 클릭한 버튼의 숫자만큼 증가한다.
- 각 파일에 각 컴포넌트가 하나씩만 있도록 (컴포넌트를) 분리해서 구현한다.

<br/>

### 📄 과제 수행 계획
- [x] <strong>1차 목표</strong>. 우선 하나의 컴포넌트에 Counter앱 기능이 제대로 동작하도록 구현
- [x] <strong>2차 목표</strong>. 하나의 파일에서 컴포넌트를 하나씩 분리하기
- [x] <strong>3차 목표</strong>. 파일을 여러 개로 나누어 분리한 컴포넌트를 하나씩 담기
- [ ] <strong>4차 목표</strong>. 리펙토링을 통한 코드 개선

<br/>

### ✏ 과제 진행
#### 1차 제출
- 우선, Counter 앱이 하나의 파일에서 정상적으로 동작하도록 구현하는 것을 1차 목표로 잡았습니다.
- App 컴포넌트 하나에 모든 Counter 앱 기능을 성공적으로 구현했습니다.  (<strong>1차 목표 달성</strong>)

 #### 2차 제출
- App 컴포넌트를 각 기능에 따라 <code>Counter</code>, <code>Button</code>, <code>Buttons</code>, <code>Page</code> 4개의 컴포넌트로 분리했습니다.  (<strong>2차 목표 달성</strong>)
- 리뷰를 참고하여, ClickMe 클릭 이벤트와 NumberButton 클릭 이벤트를 하나의 함수로 관리할 수 있도록 다시 구현했습니다.
- 각 파일에 동일한 이름의 컴포넌트를 하나씩 담아 과제를 완성했습니다. (<strong>3차 목표 달성</strong>)
- button 태그의 key 누락 문제를 해결하기 위해 key 속성을 추가했습니다.
<br/>

SangYoonLee1231 committed:
# Week 2 Assignment 1 - 'Counter 앱 만들고 파일 분리하기'

- First Submit

daadaadaah commented on src/index.jsx:
```diff
@@ -0,0 +1,43 @@
+
+  const { count } = number;
+
+  function handleClick() {
+    setNumber({ count: count + 1 });
+  }
+
+  function handleClickNumber(value) {
+    setNumber({ count: count + value });
+  }
```
1개의 함수를 사용해서 ClickMe 클릭 이벤트와 NumberButton 클릭 이벤트 둘다 관리할 수 있어요!ㅎㅎ
한번 도전해보세요~🚀

SangYoonLee1231 replied:
도전 성공했습니다~! 👊
근데 맞게 잘 한건지는 모르겠네요..ㅎㅎ 😅

daadaadaah reviewed (changes requested):
PR 내용을 정말 정성스럽게 잘 작성해주셔서 좋았습니다👍
작은 목표 하나하나씩 달성해봅시다!👊 다음 PR 기다리고 있을게요~!

SangYoonLee1231 committed:
2차 목표 달성! - 컴포넌트 분리

- App 컴포넌트를 각 기능별로 분리하여 Counter, Button, Buttons, Page 4개의 컴포넌트를 생성했습니다.

- 리뷰를 참고하여, ClickMe 클릭 이벤트와 NumberButton 클릭 이벤트를 하나의 함수로 관리할 수 있도록 다시 구현했습니다.

SangYoonLee1231 committed:
3차 목표 달성! - 각 파일에 컴포넌트 하나씩 담기

- 각 파일에 동일한 이름의 컴포넌트를 하나씩 담아 과제를 완성했습니다.

SangYoonLee1231 committed:
button 태그에 key 속성 추가

- button 태그의 key 누락 문제를 해결하기 위해 key 속성을 추가했습니다.

- 그런데 여전히 'Each child in a list should have a unique "key" prop.' 에러 문구가 나옵니다. 그 이유는 잘 모르겠습니다..

SangYoonLee1231 commented:
#### 2차 제출
- App 컴포넌트를 각 기능에 따라 <code>Counter</code>, <code>Button</code>, <code>Buttons</code>, <code>Page</code> 4개의 컴포넌트로 분리했습니다.  (<strong>2차 목표 달성</strong>)
- 리뷰를 참고하여, ClickMe 클릭 이벤트와 NumberButton 클릭 이벤트를 하나의 함수로 관리할 수 있도록 다시 구현했습니다.
- 각 파일에 동일한 이름의 컴포넌트를 하나씩 담아 과제를 완성했습니다. (<strong>3차 목표 달성</strong>)
- button 태그의 key 누락 문제를 해결하기 위해 key 속성을 추가했습니다.  
  (하지만 이렇게 해도 문제가 해결되지 않았습니다. 도움이 필요합니다!)
<br/>

SangYoonLee1231 commented on src/Buttons.jsx:
```diff
@@ -0,0 +1,14 @@
+  return (
+    <p>
+      {[1, 2, 3, 4, 5].map((i) => (
+        <Button keyNumber={i} onClick={onClick} />
```
질문이 하나 있습니다!

React 공식문서를 보니 위 Button 엘리먼트에 key 속성을 부여하는 것이 올바른 key 속성 사용법처럼 보입니다.
(key는 위 [1, 2, 3, 4, 5] 배열의 context에서만 의미가 있으므로)

하지만 실제로 <code>{[1, 2, 3, 4, 5].map((i) => \<Button key={i.toString()} onClick={onClick} /\>)}</code> 이렇게 코드를 작성해보면, Button 엘리먼트에서 해당 값을 제대로 읽지 못합니다. (console.log로 출력해보니 undefined로 값이 나옵니다.)

[에러 메세지]
react.development.js:210 Warning: Button: `key` is not a prop.
    Trying to access it will result in `undefined` being returned.
    If you need to access the same value within the child component,
    you should pass it as a different prop.

그래서 대신 위와 같이 keyNumber 속성으로 key값을 전달하여 구현을 했습니다만 이는 옳은 key 사용법처럼 보이지 않기에, 어떻게 이 문제를 해결해야 할까 싶습니다. 도움 좀 부탁드리겠습니다! 😫

SangYoonLee1231 replied:
button 태그의 key 누락 문제와 관련하여 질문이 하나 있습니다! ✋

<br/>

React 공식문서를 보니 위 코드의 8번째 줄에서 Button 엘리먼트에 key 속성을 부여하는 것이 올바른 key 속성 사용법처럼 보입니다.
(key는 위 [1, 2, 3, 4, 5] 배열의 context에서만 의미가 있으므로)
<a href="https://ko.reactjs.org/docs/lists-and-keys.html#extracting-components-with-keys
">참고한 공식문서 해당 위치</a> 

하지만 실제로 <code>{[1, 2, 3, 4, 5].map((i) => \<Button key={i.toString()} onClick={onClick} /\>)}</code> 이렇게 코드를 작성해보면, Button 엘리먼트에서 해당 값을 제대로 읽지 못합니다. (console.log로 출력해보니 undefined로 값이 나옵니다.)

[에러 메세지]
react.development.js:210 Warning: Button: `key` is not a prop.
    Trying to access it will result in `undefined` being returned.
    If you need to access the same value within the child component,
    you should pass it as a different prop.

그래서 대신 위와 같이 keyNumber 속성으로 key값을 전달하여 구현을 했습니다만 이는 옳은 key 사용법처럼 보이지 않기에, 어떻게 이 문제를 해결해야 할까 싶습니다. 도움 좀 부탁드리겠습니다! 😫

daadaadaah commented on src/App.jsx:
```diff
@@ -0,0 +1,27 @@
+  const { count } = number;
+
+  function handleClick(value) {
+    if (!value) {
+      setNumber({ count: count + 1 });
+      return;
+    }
```
if문 없이도 구현할 수 있습니다! 헌번 도전해보세요!ㅎㅎ

daadaadaah commented on src/Buttons.jsx:
```diff
@@ -0,0 +1,14 @@
+function Buttons({ onClick }) {
+  return (
+    <p>
+      {[1, 2, 3, 4, 5].map((i) => (
+        <Button keyNumber={i} onClick={onClick} />
+      ))}
```
리액트에서 `map()` 같이 반복문을 이용해서 여러개의 값을  출력할 때는 컴포넌트 속성에 반드시 `key` 를 추가해야 합니다. 왜 key가 필요한지에 대해 한번 공부해보세요!
참고 : [https://reactjs.org/docs/lists-and-keys.html](https://reactjs.org/docs/lists-and-keys.html)

SangYoonLee1231 replied:
감사합니다! key에 대해 좀 더 공부해보겠습니다!

daadaadaah reviewed (changes requested):
3차 목표까지 달성하셨네요! 잘하셨어요~ 👏
key 속성 관련해서 링크 하나 첨부했습니다! 한번 공부해보세요~ㅎㅎ
